### PR TITLE
Fixes #11211: improve the usability of the parameter expander.

### PR DESCRIPTION
**Pull request by waldenraines**
http://projects.theforeman.org/issues/11211

### CODE REVIEW

**Comment by waldenraines** [comment]
Please let me know if there are other expanders that need replacing that I missed.

**Comment by tbrisker** [comment]
@waldenraines right now it appears on on global parameters, not for class parameters, smart variables and overrides

File: app/helpers/common_parameters_helper.rb [hunk 1]
```diff
@@ -22,4 +22,10 @@ def parameter_value_field(value)
   def use_puppet_default_help link_title = _("Explain use Puppet default"), title = _("Use Puppet default")
     popover(link_title, _("Do not send this parameter via the ENC.<br>Puppet will use the value defined in the manifest."), :title => title)
   end
+
+  def include_javascript
```
Comment (ohadlevy) [inline app/helpers/common_parameters_helper.rb]: can the name relate to the editor? plain javascript is too generic.

File: app/helpers/provisioning_templates_helper.rb [hunk 2]
```diff
@@ -13,7 +13,7 @@ def template_kind(template)
   end
 
   def include_javascript
-    javascript 'provisioning_template', 'ace/ace',
+    javascript 'editor', 'ace/ace',
```
Comment (ohadlevy) [inline app/helpers/provisioning_templates_helper.rb]: left over?

Comment (waldenraines) [inline app/helpers/provisioning_templates_helper.rb]: What do you mean?

Comment (domcleal) [inline app/helpers/provisioning_templates_helper.rb]: Perhaps he was expecting this method to have been removed, given the one above - however the one above is only in common_parameters_helper.  I'd suggest moving it to one of the app-wide helpers or a new EditorHelper (renaming it to something less generic in the process), then delete the duplicates in these two.

**Comment by domcleal** [comment]
> @waldenraines right now it appears on on global parameters, not for class parameters, smart variables and overrides

I'm not sure how that would work, it's a huge UI to add in the middle of those forms.  It kind of works here because it's a separate form.

1) Javascript needs including on common_parameters/edit too, else it only works in two-pane.
2) assets list in config/environments/production.rb needs updating for the filename change

**Comment by waldenraines** [comment]
> I'm not sure how that would work, it's a huge UI to add in the middle of those forms. It kind of works here because it's a separate form.

Can you please elaborate on why you think it wouldn't work?  I pushed new code that has the editor on these pages so you can see if you like it.

**Comment by waldenraines** [comment]
> 1) Javascript needs including on common_parameters/edit too, else it only works in two-pane.
> 2) assets list in config/environments/production.rb needs updating for the filename change

Fixed these issues.

**Comment by orrabin** [comment]
@waldenraines I agree with @domcleal on class parameters - the editor there should either be opened only in full screen mode or adapt to the size of the text since right now adding it to parameters with type boolean looks like this:
image

and that is too much noise for a simple true\false value.
This will be even more problematic when trying to merge https://github.com/theforeman/foreman/pull/2398 as the matchers have moved into a table:
image

File: app/assets/javascripts/editor.js [hunk 3]
```diff
@@ -85,45 +85,50 @@ function snippet_changed(item){
   $('#association').toggle(!checked);
 }
 
-function create_editor(item) {
-  item.parent().prepend("<div id='editor1'></div>");
-  item.hide();
-  Editor = ace.edit("editor1");
+function create_editor() {
+  var editorId = "editor" + Math.floor((Math.random() * 100) + 1),
```
Comment (domcleal) [inline app/assets/javascripts/editor.js]: This looks like it could random()ly fail too, some sort of ID collision avoidance?

**Comment by domcleal** [comment]
And yeah, as @orrabin said, it's as I suspected.. the Puppet class edit form has become enormous inside, I don't think it's usable.

File: app/assets/javascripts/editor.js (hunk 3 above)
Comment (tbrisker) [inline app/assets/javascripts/editor.js]: According to the birthday paradox, this will collide with 50% probability on 10 editors.

Comment (waldenraines) [inline app/assets/javascripts/editor.js]: Any suggestions on what to do here?  I could add some more zeros.

I'd rewrite this whole thing in AngularJS if I thought it would have a chance of being merged.  Then we wouldn't need to add this unique ID nonsense and it would be tested.  But alas.

**Comment by waldenraines** [comment]
> @waldenraines I agree with @domcleal on class parameters - the editor there should either be opened only in full screen mode or adapt to the size of the text since right now adding it to parameters with type boolean looks like this:

What about adjusting the size based on the type of parameter?

I could also adjust the height of it in general such that it's shorter.

@domcleal do keep this in mind:  http://uxmyths.com/post/654047943/myth-people-dont-scroll

File: app/assets/javascripts/editor.js (hunk 3 above)
Comment (ohadlevy) [inline app/assets/javascripts/editor.js]: On Wed, Aug 5, 2015 at 7:14 PM, Walden Raines notifications@github.com
wrote:

I'd rewrite this whole thing in AngularJS if I thought it would have a

> chance to be merged. Then we wouldn't need to add this unique ID nonsense
> and it would be tested. But alas.

What would it mean to write it in Angular? /me personally don't object
better ways to solve problems? Would it require the entire bastion plugin
to support it?

Comment (waldenraines) [inline app/assets/javascripts/editor.js]: > What would it mean to write it in Angular? /me personally don't object
> better ways to solve problems? Would it require the entire bastion plugin
> to support it?

No, any html element can be an angular "app" as long as the angular javascript is included.  So these pages themselves could be an angular "app"s just so they could use the directive I would create for this editor.

As far as including the Angular JS source goes, we wouldn't necessarily have to include bastion in order to get this to work.  A stop-gap could be to include the angular source independently, we could even use something like [angularjs-rails](https://github.com/hiravgandhi/angularjs-rails).

Comment (domcleal) [inline app/assets/javascripts/editor.js]: Where is there the chance of collisions?  As far as I can tell, it would only be on the class edit form, which as per the other comments, is a bad idea anyway.

**Comment by domcleal** [comment]
> What about adjusting the size based on the type of parameter?
> 
> I could also adjust the height of it in general such that it's shorter.

I don't think that it is suitable to put in that location, especially when you consider #2398.

**Comment by waldenraines** [comment]
> I don't think that it is suitable to put in that location, especially when you consider #2398.

So showing a tiny and unusable single line text area is better?

screen shot 2015-08-06 at 10 19 36 am

Why isn't putting the editor here suitable, assuming I change it such that its size is adjusted based on its contents?

**Comment by domcleal** [comment]
You've got both a full screen button and a resize control available on that text field, seems enough?

File: app/assets/javascripts/editor.js (hunk 3 above)
Comment (tbrisker) [inline app/assets/javascripts/editor.js]: /me thinks angular is major overkill for giving a unique id for a small input field.

Comment (waldenraines) [inline app/assets/javascripts/editor.js]: The benefit is not just unique IDs, there are also major benefits in maintainability, readability, reusability, and testability but this argument is out of scope for this PR.

**Comment by waldenraines** [comment]
@domcleal I can certainly remove the editor from this page.  I was asked to add it to this page by @tbrisker.  Can you two please reach a consensus on what is desired?  I will go with whatever y'all decide as you two are more familiar with how these pages are used.

**Comment by waldenraines** [comment]
@domcleal, @tbrisker, @orrabin removed the editor from the class parameters.  Please let me know what else you would like me to do.

**Comment by domcleal** [comment]
Before I review much further, can I just check precisely what the ticket issue is?  It states that "the parameter expander currently does not maintain formatting" - this is the full screen box on the global parameters form?  It seems that it does, and the same full screen expansion is used in other places, which keeps formatting (except that YAML/JSON gets reformatted in lookup values).

**Comment by waldenraines** [comment]
@orrabin pushed update that allows the selection of ace editor mode.

> Before I review much further, can I just check precisely what the ticket issue is? It states that "the parameter expander currently does not maintain formatting" - this is the full screen box on the global parameters form? It seems that it does, and the same full screen expansion is used in other places, which keeps formatting (except that YAML/JSON gets reformatted in lookup values).

Part of the issue was that cram multiline values into a single line text area.  The thought was to allow users to edit values without having to go full screen.

**Comment by tbrisker** [comment]
I agree that this looks bad on the smart parameters forms and it's good that you removed it (sorry for the delayed response). I tend to think that the full-screen button and the text-field size control are enough for handling longer values, but others may disagree. 
This PR is good in my opinion mostly because it decouples the ace editor from the template form, making it a more modular piece that can be easily reused in other parts of Foreman (or plugins etc.). 
It doesn't solve the original issue of handling longer parameter values better, but I'm not sure if there is a good way to solve it without sacrificing the clean layout of the page.

**Comment by orrabin** [comment]
@waldenraines is there a way to show the editor only when in full screen?
I mean when you are editing a smart parameter and your not in full screen leave it as is and when opening it as full screen you get the editor.

**Comment by waldenraines** [comment]
> @waldenraines is there a way to show the editor only when in full screen?
> I mean when you are editing a smart parameter and your not in full screen leave it as is and when opening it as full screen you get the editor.

For all of the editors?  I thought we only wanted to do this for class parameters.

**Comment by orrabin** [comment]
@waldenraines I meant in general is it possible? If it is then that is the best solution for class parameters.

**Comment by dLobatog** [comment]
This needs rebasing. After a lot of reading, I couldn't really figure out what would be the outcome of this PR, if it's going to end up as a refactor maybe submit a different ticket. A couple of screenshots highlighting the changes make it easier to review too :smile:

**Comment by dLobatog** [comment]
@waldenraines any updates?

**Comment by waldenraines** [comment]
> @waldenraines any updates?

@eLobato It doesn't seem like we don't have agreement on what is wanted here.  

The original goal was to have an "expander" that maintains the formatting of the pasted in value after saving and formats the value as you type.  In order to accomplish these aims we wanted to use the existing ace editor rather than introducing a new library.

The feedback was that the editor within the form is undesirable on the puppet class parameters page and that it should appear when the user clicks the full screen button.  So if we can all agree on that I'm happy to try to make that work.  I'm also happy to close this and work on something else instead.  I just want to avoid spending time on changes that aren't even wanted in the first place.

**Comment by waldenraines** [comment]
@domcleal @eLobato @orrabin 

In order to get this to work on any page that has more than one editor (i.e puppet class parameters) it will take a fair amount of refactoring because the editor relies on there only being one editor on the page.  

How would y'all like me to proceed?  I could close this and reopen another PR with the changes in this PR against a newly created "refactor" issue and then attempt to make the changes necessary to make this work as a "Fixes #11211" issue.

Thoughts?

**Comment by waldenraines** [comment]
Closing in favor of #2656.
